File: life-cpp/board.cpp
#include "board.h"

#include <iostream>

stdx::optional<bool> Grid::at(ssize_t x, ssize_t y) const
{
	auto idx = index(x, y);
	if(idx != stdx::nullopt)
	{
		return m_grid[*idx];
	}
	return stdx::nullopt;
}

stdx::optional<size_t> Grid::index(ssize_t x, ssize_t y) const
{
	if(x < 0 || x >= width())
	{
		return stdx::nullopt;
	}

	if(y < 0 || y >= height())
	{
		return stdx::nullopt;
	}

	return stdx::optional<size_t>(y * height() + x);
}

void Grid::set(ssize_t x, ssize_t y, bool val)
{
	auto idx = index(x, y);
	if(idx != stdx::nullopt)
	{
		m_grid[*idx] = val;
	}
}

std::array<stdx::optional<size_t>, 8> Grid::neighbors(ssize_t x, ssize_t y) const
{
	return {
		index(x - 1, y - 1),
		index(x - 1, y),
		index(x - 1, y + 1),
		index(x, y - 1),
		index(x, y + 1),
		index(x + 1, y - 1),
		index(x + 1, y),
		index(x + 1, y + 1)
	};
}

size_t Grid::count_live_neighbors(ssize_t x, ssize_t y) const
{
	size_t count = 0;

	for(auto idx : neighbors(x, y))
	{
		if(idx != stdx::nullopt)
		{
			count += m_grid[*idx];
		}
	}

	return count;
}

bool LifeBoard::will_live(bool is_live, size_t num_neighbors)
{
	if(is_live)
	{
		return num_neighbors == 2 || num_neighbors == 3;
	}
	else
	{
		return num_neighbors == 3;
	}
}

void LifeBoard::update()
{
	for(size_t y = 0; y < m_current.height(); ++y)
	{
		for(size_t x = 0; x < m_current.width(); ++x)
		{
			bool is_live = m_current.at(x, y).value_or(false);
			size_t num_neighbors = m_current.count_live_neighbors(x, y);

			m_next.set(x, y, will_live(is_live, num_neighbors));
		}
	}
}

const Grid *LifeBoard::tick()
{
	std::cout << "C++ board tick" << std::endl;
	update();
	std::swap(m_current, m_next);

	return &m_current;
}

File: life-rs/src/board.rs
use crate::traits::{Grid, LifeBoard};
use itertools::Itertools;

pub struct NativeGrid(Vec<Vec<bool>>);

impl NativeGrid {
    fn get_neighbors(&self, x: isize, y: isize) -> [(isize, isize); 8] {
        [
            (x - 1, y - 1),
            (x - 1, y),
            (x - 1, y + 1),
            (x, y - 1),
            (x, y + 1),
            (x + 1, y - 1),
            (x + 1, y),
            (x + 1, y + 1),
        ]
    }

    fn iter_coords(&self) -> impl Iterator<Item = (isize, isize)> {
        (0..self.width() as isize).cartesian_product(0..self.height() as isize)
    }
}

impl Grid for NativeGrid {
    fn new(width: usize, height: usize) -> NativeGrid {
        NativeGrid(vec![vec![false; width]; height])
    }

    fn width(&self) -> usize {
        self.0[0].len()
    }

    fn height(&self) -> usize {
        self.0.len()
    }

    fn at(&self, x: isize, y: isize) -> Option<bool> {
        self.0
            .get(y as usize)
            .and_then(|row| row.get(x as usize))
            .map(|val| *val)
    }

    fn set(&mut self, x: isize, y: isize, val: bool) {
        self.0
            .get_mut(y as usize)
            .and_then(|row| row.get_mut(x as usize))
            .map(|cell| *cell = val);
    }

    fn count_live_neighbors(&self, x: isize, y: isize) -> usize {
        self.get_neighbors(x, y)
            .iter()
            .map(|(x, y)| self.at(*x, *y).unwrap_or_default() as usize)
            .sum()
    }
}

pub struct NativeLifeBoard {
    current: NativeGrid,
    next: NativeGrid,
}

impl NativeLifeBoard {
    fn will_live(is_live: bool, num_live_neighbors: usize) -> bool {
        if is_live {
            num_live_neighbors == 2 || num_live_neighbors == 3
        } else {
            num_live_neighbors == 3
        }
    }

    fn update(&mut self) {
        for (x, y) in self.current.iter_coords() {
            let is_live = self.current.at(x, y).unwrap_or_default();
            let num_live_neighbors = self.current.count_live_neighbors(x, y);
            self.next
                .set(x, y, Self::will_live(is_live, num_live_neighbors));
        }
    }
}

impl LifeBoard for NativeLifeBoard {
    type Grid = NativeGrid;

    fn new(initial: NativeGrid) -> Self {
        let width = initial.width();
        let height = initial.height();

        NativeLifeBoard {
            current: initial,
            next: NativeGrid::new(width, height),
        }
    }

    fn tick(&mut self) -> &NativeGrid {
        println!("Rust board tick");
        self.update();
        std::mem::swap(&mut self.current, &mut self.next);
        &self.current
    }
}
